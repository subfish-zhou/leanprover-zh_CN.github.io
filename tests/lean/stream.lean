def g (xs : Array Nat) (ys : Array Nat) : IO Unit := do
  let mut s := toStream ys
  for x in xs do
    match Stream.next? s with
    | none => break
    | some (y, s') =>
      s := s'
      IO.println s!"x: {x}, y: {y}"

#eval g #[1, 2, 3] #[4, 5]
#print "-----"

def f [Stream ρ α] [ToString α] (xs : Array Nat) (ys : ρ) : IO Unit := do
  let mut ys := ys
  for x in xs do
    match Stream.next? ys with
    | none => break
    | some (y, ys') =>
      ys := ys'
      IO.println s!"x: {x}, y: {y}"

#eval f #[1, 2, 3] (toStream #[4, 5])
#print "-----"
#eval f #[1, 2] (toStream [4, 5, 6])
#print "-----"
#eval f #[1, 2, 3] (toStream "hello")

def h [Stream ρ α] [ToString α] (s : ρ) : IO Unit := do
  for a in s do
    IO.println a

#eval h (toStream [1, 2, 3])
#print "-----"
#eval h [1, 2, 3]
#print "-----"
#eval h (toStream #[1, 2, 3])
#print "-----"
#eval h #[1, 2, 3, 4, 5, 6, 7][2:5]
#print "-----"
#eval h ([1, 2, 3], [4, 5, 6])
