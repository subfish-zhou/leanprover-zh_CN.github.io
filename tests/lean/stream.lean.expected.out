x: 1, y: 4
x: 2, y: 5
-----
x: 1, y: 4
x: 2, y: 5
-----
x: 1, y: 4
x: 2, y: 5
-----
x: 1, y: h
x: 2, y: e
x: 3, y: l
1
2
3
-----
1
2
3
-----
1
2
3
-----
3
4
5
-----
(1, 4)
(2, 5)
(3, 6)
